pkg/migration/plan_generator.go

```diff
@@ -0,0 +1,501 @@
+	if _, ok, err := toManagedResource(u.Object); err != nil || !ok {
+		// not a managed resource or unable to determine
+		// whether it's a managed resource
+		return nil // nolint:nilerr
+	}
+	qName := getQualifiedName(u.Object)
+	if err := pg.stepPauseManagedResource(u, qName); err != nil {
+		return err
+	}
+	if err := pg.stepOrphanManagedResource(u, qName); err != nil {
+		return err
+	}
+	pg.stepDeleteOldManagedResource(u)
+	return nil
+}
+
+func (pg *PlanGenerator) stepStartManagedResource(u *UnstructuredWithMetadata) error {
+	annot := u.Object.GetAnnotations()
+	if annot != nil {
+		delete(annot, meta.AnnotationKeyReconciliationPaused)
+		u.Object.SetAnnotations(annot)
+	}
+
+	u.Metadata.Path = fmt.Sprintf("%s/%s.yaml", pg.Plan.Spec.Steps[stepStartManaged].Name, getQualifiedName(u.Object))
+	pg.Plan.Spec.Steps[stepStartManaged].Apply.Files = append(pg.Plan.Spec.Steps[stepStartManaged].Apply.Files, u.Metadata.Path)
+	if err := pg.target.Put(*u); err != nil {
+		return errors.Wrap(err, errResourceOutput)
+	}
+	return nil
+}
+
+func (pg *PlanGenerator) stepPauseManagedResource(u *UnstructuredWithMetadata, qName string) error {
+	u.Metadata.Path = fmt.Sprintf("%s/%s.yaml", pg.Plan.Spec.Steps[stepPauseManaged].Name, qName)
+	pg.Plan.Spec.Steps[stepPauseManaged].Apply.Files = append(pg.Plan.Spec.Steps[stepPauseManaged].Apply.Files, u.Metadata.Path)
+	return pg.pause(u.Metadata.Path, &u.Object)
+}
+
+func (pg *PlanGenerator) stepPauseComposite(u *UnstructuredWithMetadata) error {
+	u.Metadata.Path = fmt.Sprintf("%s/%s.yaml", pg.Plan.Spec.Steps[stepPauseComposites].Name, getQualifiedName(u.Object))
+	pg.Plan.Spec.Steps[stepPauseComposites].Apply.Files = append(pg.Plan.Spec.Steps[stepPauseComposites].Apply.Files, u.Metadata.Path)
+	return pg.pause(u.Metadata.Path, &u.Object)
+}
+
+func (pg *PlanGenerator) stepOrphanManagedResource(u *UnstructuredWithMetadata, qName string) error {
+	pv := fieldpath.Pave(u.Object.Object)
+	if err := pv.SetValue("spec.deletionPolicy", v1.DeletionOrphan); err != nil {
+		return errors.Wrap(err, errDeletionOrphan)
+	}
+	u.Metadata.Path = fmt.Sprintf("%s/%s.yaml", pg.Plan.Spec.Steps[stepDeletionPolicyOrphan].Name, qName)
+	pg.Plan.Spec.Steps[stepDeletionPolicyOrphan].Apply.Files = append(pg.Plan.Spec.Steps[stepDeletionPolicyOrphan].Apply.Files, u.Metadata.Path)
+	return errors.Wrap(pg.target.Put(*u), errResourceOrphan)
+}
+
+func (pg *PlanGenerator) stepDeleteOldManagedResource(u *UnstructuredWithMetadata) {
+	pg.Plan.Spec.Steps[stepDeleteOldManaged].Delete.Resources = append(pg.Plan.Spec.Steps[stepDeleteOldManaged].Delete.Resources,
+		Resource{
+			GroupVersionKind: FromGroupVersionKind(u.Object.GroupVersionKind()),
+			Name:             u.Object.GetName(),
+		})
+}
+
+func addPauseAnnotation(u *unstructured.Unstructured) {
+	annot := u.GetAnnotations()
+	if annot == nil {
+		annot = make(map[string]string)
+	}
+	annot[meta.AnnotationKeyReconciliationPaused] = "true"
+	u.SetAnnotations(annot)
```

Comment:
nit: There is `meta.AddAnnotations` function in runtime that would help here.

Reply:
Done.